docs/csiaddonsjob.md

```diff
@@ -0,0 +1,44 @@
+# CSIAddonsJob
+
+CSIAddonsJob is a namespaced custom resource designed to invoke operation to be performed on a volume.
+
+```yaml
+apiVersion: csiaddons.openshift.io/v1beta1
+kind: CSIAddonsJob
+metadata:
+  name: sample-1
+spec:
+  operation:
+    reclaimSpace:
```

Comment:
can this be a generic one as below? if we keep it generic one we don't need to introduce a new CR for every operation we support.

```yaml
operation: reclaimSpace
target:
    pv: pvc-1232-4255-5675
parameters:
   key1: value1
   key2: value2
```

Reply:
>can this be a generic one as below? if we keep it generic one we don't need to introduce a new CR for every operation we support.

Previous iteration had similar design, it can be found here https://github.com/Rakshith-R/kubernetes-csi-addons/blob/api-job-generic/docs/csiaddonsjob.md . 
But @nixpanic preferred strongly typed CR 
refer discussion :https://github.com/csi-addons/kubernetes-csi-addons/pull/4#discussion_r738982807

Reply:
as we add more operations this will get extended a lot?

Reply:
yes, we'll extend this for each operation. Every operation may have different parameters. Maybe we want to add a `fsck` operation, and that may well need a `dry-run`, `force` or `clone-as-backup` option that is not applicable for `reclaimSpace`.

Reply:
as the parameters are opaque the user can specify the parameter as per the operations? so that a single operation can be reused instead of extending the spec?

Reply:
That makes it more difficult to validate, and generated documentation could easily be incomplete. With a strongly typed CRD, it is very clear what operations are available, and what options they accept.

Reply:
Also, from my understanding, it is not an issue to extend a CRD, as long as the existing members/values stay valid. Therefore, it should not be problematic to add more operations later.

Reply:
I prefer generic approach too,IMO:

- CSIAddonsJob CR will just be a yaml/json config file which the user has to write by referring the **documentation**(which will be responsible for clearly stating the support for operations and their options). 
- Validation may inturn be in form of having 
  - enums set on the fields (like operation: <supported_types> where only 'reclaimSpace','fsck' will be accepted).  
  - admission webhooks which reject at CR create request with clear messages and which can be easily programmed.
  - At the controller level, through error messages(this can be driver specific validation, while the other two will be generic). 
- I prefer the CR & CRD to be simple and the need to extend/change it to be very minimal. 
  

With strongly typed CR,CRD: 

- The need of change will be very often, with change in very operation/ option.
- The code for the api may indeed become large and confusing(very low reusability)
- The documentation may get confusing too.
- versioning of the api may get complicated too (?).

Reply:
Agree with the above points that what I have in my mind to suggest is to keep it generic instead of strongly typed CR's. IMHO it may look simple for now but when we start adding more operations it becomes complex.

Reply:
For each operation we need to extend things somewhere. It is just that the location(s) where the extensions need to be done should be kept as closely together (ideally a single place).

I prefer to extend things at a location so that other parts can use the (re)generated results. If the CRD can be validated easily, the admission webhook would mostly be using generated code. The same counts for (generated) documentation of the valid operations and their parameters. Anything that is done should work together with [Server-Side Apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) as well.

Requiring users to review the documentation which contains a description like a switch-statement, depending on the (string :fearful:?) value that they provide as an argument, will be very user unfriendly.

Remember that we expect users (persons, not automation) to create the CR. Input validation is where a lot of bugs are  happening, and reducing that from occurring is important. Any validation that we manually need to write has a high chance on containing issues, whereas a strong-typed automatically generated validation should have fewer problems.

For me, the interactions that the user needs to do are most important. Users will be able to make less mistakes when the CRD is strongly typed. Recent versions of `kubectl` give reasonable errors when the CR contains invalid values. This uses standard Kubernetes mechanisms, just like any other object that the user may create. The common usage, standard error reporting and re-using existing functions outweighs the required updating of the CRD (which would not happen *that* often once common storage maintenance operations are included) IMHO.

Reply:
> That makes it more difficult to validate, and generated documentation could easily be incomplete. With a strongly typed CRD, it is very clear what operations are available, and what options they accept.

If the intent is to make strongly-typed CRDs, then why are we naming this with such a generic name: "CSIAddonsJob"?

IMO, if we want strongly-typed, then changing this to CSIReclaimSpaceAddonJob would be better, and we should create a new type of job for each operation. Having a generic job with strongly-typed internals is making two different design decisions in different ways for the same resource.

For me, I think `CSIAddonsJob` is a useful generic, and generics should not be strongly-typed. I consider this somewhat analogous to a StorageClass, which is also generic and takes all kinds of opaque and untyped data.

If we need immediate/good kubectl feedback about it, a validating webhook can be used.

Reply:
>Having a generic job with strongly-typed internals is making two different design decisions in different ways for the same resource.

I agree with the above points mentioned by Blaine. 
IMO too we should either have a:
-  generic `CSIAddonsJob` (My preference)
- or Strongly typed Specific `CSIReclaimSpaceAddonJob` type(and add other jobs following similar pattern but with different type + controller)

@nixpanic wdyt?